neuro_san/service/mcp/processors/mcp_tools_processor.py

```diff
@@ -91,15 +97,36 @@ async def call_tool(self, request_id, metadata: Dict[str, Any], tool_name: str,
         input_request: Dict[str, Any] = self._get_chat_input_request(prompt)
         response_text: str = ""
         try:
-            result_generator = service.streaming_chat(input_request, metadata)
-            async for result_dict in result_generator:
-                partial_response: str = await self._extract_tool_response_part(result_dict)
-                if partial_response is not None:
-                    response_text = response_text + partial_response
+            async with asyncio.timeout(self.tool_timeout_seconds):
+                result_generator = service.streaming_chat(input_request, metadata)
+                async for result_dict in result_generator:
+                    partial_response: str = await self._extract_tool_response_part(result_dict)
+                    if partial_response is not None:
+                        response_text = response_text + partial_response
+
+        except (asyncio.CancelledError, tornado.iostream.StreamClosedError):
+            self.logger.info(metadata, "Tool execution %s cancelled/stream closed.", tool_name)
+            return McpErrorsUtil.get_tool_error(request_id, f"Stream closed for tool {tool_name}")
+
+        except asyncio.TimeoutError:
+            self.logger.info(metadata,
+                             "Chat tool timeout for %s in %f seconds.",
+                             tool_name, self.tool_timeout_seconds)
+            return McpErrorsUtil.get_tool_error(request_id, f"Timeout for tool {tool_name}")
+
         except Exception as exc:  # pylint: disable=broad-exception-caught
             self.logger.error(metadata, "Tool %s execution failed: %s", tool_name, str(exc))
             return McpErrorsUtil.get_tool_error(request_id, f"Failed to execute tool {tool_name}")
 
+        finally:
+            # We are done with response stream,
+            # ensure generator is closed properly in any case:
+            if result_generator is not None:
+                with contextlib.suppress(Exception):
+                    # It is possible we will call .aclose() twice
+                    # on our result_generator - it is allowed and has no effect.
+                    await result_generator.aclose()
+
```

Comment:
Simplified code a little - we do our generator closing in a single place of "finally" block.
For each specific code path (some exception or normal execution) we only do specific required processing.